File: src/build_slides.py
import json
import os
import sys

def build_slides(episode_dir):
    source_path = os.path.join(episode_dir, "草稿1.md")
    # もし 草稿1.md がなければ フォルダ内の唯一の .md を探す
    if not os.path.exists(source_path):
        md_files = [f for f in os.listdir(episode_dir) if f.endswith('.md')]
        if md_files:
            source_path = os.path.join(episode_dir, md_files[0])
        else:
            print(f"Error: No .md manuscript found in {episode_dir}")
            return

    manifest_path = os.path.join(episode_dir, "manifest.json")
    output_path = os.path.join(episode_dir, "slides.js")

    if not os.path.exists(manifest_path):
        print(f"Error: {manifest_path} not found.")
        return

    # Load Manifest
    with open(manifest_path, 'r', encoding='utf-8') as f:
        manifest = json.load(f)

    # Load Source File
    with open(source_path, 'r', encoding='utf-8') as f:
        source_lines = f.readlines()

    final_slides = []
    for entry in manifest.get('slides', []):
        line_range = entry.get('range')
        if not line_range or len(line_range) != 2:
            continue
        
        start_line = line_range[0]
        end_line = line_range[1]
        selected_lines = source_lines[start_line-1 : end_line]
        verbatim_script = "".join(selected_lines).strip()

        slide = {
            "title": entry.get("title", ""),
            "content": entry.get("content", ""),
            "imgDesc": entry.get("imgDesc", ""),
            "script": verbatim_script
        }
        
        # もし画像が既にある場合はそれを使う設定（将来用）
        # imageSrc はフォルダ相対パスで管理
        img_filename = entry.get("imageSrc", "")
        if img_filename:
            slide["imageSrc"] = img_filename
        
        final_slides.append(slide)

    # Write slides.js
    with open(output_path, 'w', encoding='utf-8') as f:
        f.write("/**\n * HYBRID BUILD: AI Design + Algorithmic Verbatim Transfer\n")
        f.write(" * Source: " + os.path.basename(source_path) + "\n")
        f.write(" */\n")
        f.write("const slidesData = ")
        json.dump(final_slides, f, ensure_ascii=False, indent=4)
        f.write(";")

    print(f"Success: Built {len(final_slides)} slides in {output_path}")

if __name__ == "__main__":
    if len(sys.argv) > 1:
        target_dir = sys.argv[1]
    else:
        # 引数がない場合は、現在の直下の episodes から選択するなどの簡易対話を想定
        episodes_dir = "episodes"
        if not os.path.exists(episodes_dir):
            episodes_dir = "../episodes" # srcから実行する場合

        available = sorted([d for d in os.listdir(episodes_dir) if os.path.isdir(os.path.join(episodes_dir, d))])
        
        if not available:
            print("No episodes found in episodes/ directory.")
            sys.exit(1)
            
        print("Available Episodes:")
        for i, name in enumerate(available):
            print(f"[{i}] {name}")
        
        idx = input("Select episode index to build: ")
        try:
            target_dir = os.path.join(episodes_dir, available[int(idx)])
        except:
            print("Invalid selection.")
            sys.exit(1)

    build_slides(target_dir)

File: src/build_slides.mjs
#!/usr/bin/env node
/**
 * manifest.json + Markdown 原稿 → slides.js（ビルド専用・依存パッケージなし）
 */
import fs from "node:fs";
import path from "node:path";
import readline from "node:readline/promises";
import { stdin as input, stdout as output } from "node:process";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const repoRoot = path.resolve(__dirname, "..");

/** 原稿を行配列に分ける（改行を行末に保持。manifest の 1 始まり range と整合） */
function readlinesUtf8(filePath) {
    const normalized = fs.readFileSync(filePath, "utf8").replace(/\r\n/g, "\n");
    if (normalized.length === 0) return [];
    const lines = [];
    let pos = 0;
    while (pos < normalized.length) {
        const nl = normalized.indexOf("\n", pos);
        if (nl === -1) {
            lines.push(normalized.slice(pos));
            break;
        }
        lines.push(normalized.slice(pos, nl + 1));
        pos = nl + 1;
    }
    return lines;
}

function findEpisodeDirs(episodesRoot) {
    const results = [];
    function walk(dir) {
        if (!fs.existsSync(dir)) return;
        if (fs.existsSync(path.join(dir, "manifest.json"))) {
            results.push(dir);
            return;
        }
        for (const name of fs.readdirSync(dir, { withFileTypes: true })) {
            if (name.isDirectory()) walk(path.join(dir, name.name));
        }
    }
    walk(episodesRoot);
    return results.sort();
}

function buildSlides(episodeDir) {
    const absDir = path.isAbsolute(episodeDir)
        ? episodeDir
        : path.resolve(process.cwd(), episodeDir);

    let sourcePath = path.join(absDir, "草稿1.md");
    if (!fs.existsSync(sourcePath)) {
        const mdFiles = fs
            .readdirSync(absDir)
            .filter((f) => f.endsWith(".md"))
            .map((f) => path.join(absDir, f));
        if (mdFiles.length === 0) {
            console.error(`Error: No .md manuscript found in ${absDir}`);
            process.exitCode = 1;
            return;
        }
        sourcePath = mdFiles[0];
    }

    const manifestPath = path.join(absDir, "manifest.json");
    const outputPath = path.join(absDir, "slides.js");

    if (!fs.existsSync(manifestPath)) {
        console.error(`Error: ${manifestPath} not found.`);
        process.exitCode = 1;
        return;
    }

    const manifest = JSON.parse(fs.readFileSync(manifestPath, "utf8"));
    const sourceLines = readlinesUtf8(sourcePath);

    const finalSlides = [];
    for (const entry of manifest.slides || []) {
        const lineRange = entry.range;
        if (!lineRange || lineRange.length !== 2) continue;

        const startLine = lineRange[0];
        const endLine = lineRange[1];
        const selectedLines = sourceLines.slice(startLine - 1, endLine);
        const verbatimScript = selectedLines.join("").trim();

        const slide = {
            title: entry.title ?? "",
            content: entry.content ?? "",
            imgDesc: entry.imgDesc ?? "",
            script: verbatimScript,
        };
        if (entry.imageSrc) slide.imageSrc = entry.imageSrc;
        finalSlides.push(slide);
    }

    const header =
        "/**\n" +
        " * HYBRID BUILD: AI Design + Algorithmic Verbatim Transfer\n" +
        ` * Source: ${path.basename(sourcePath)}\n` +
        " */\n" +
        "var slidesData = ";
    const body = JSON.stringify(finalSlides, null, 4) + ";";
    fs.writeFileSync(outputPath, header + body, "utf8");

    console.log(`Success: Built ${finalSlides.length} slides in ${outputPath}`);
}

const argv = process.argv.slice(2);

if (argv[0]) {
    buildSlides(argv[0]);
} else {
    const episodesDir = fs.existsSync(path.join(process.cwd(), "episodes"))
        ? path.join(process.cwd(), "episodes")
        : path.join(repoRoot, "episodes");

    const available = findEpisodeDirs(episodesDir);
    if (available.length === 0) {
        console.error("No episodes with manifest.json found under", episodesDir);
        process.exit(1);
    }

    console.log("Available episodes (manifest.json):");
    available.forEach((dir, i) => {
        console.log(`[${i}] ${path.relative(repoRoot, dir) || "."}`);
    });

    const rl = readline.createInterface({ input, output });
    const idxStr = await rl.question("Select episode index to build: ");
    rl.close();

    const idx = parseInt(idxStr, 10);
    if (Number.isNaN(idx) || idx < 0 || idx >= available.length) {
        console.error("Invalid selection.");
        process.exit(1);
    }

    buildSlides(available[idx]);
}
